### ISSUE TITLE
What does load-path do?

### ISSUE THREAD

**Issue by kshenoy** [jwiegley/use-package#261]
From the [README](https://github.com/jwiegley/use-package#extending-the-load-path), it seems like `:load-path` can be used to specify any additional dirs that are required. Also from your [video demonstration](https://youtu.be/2TSKxxYEbII?list=WL&t=252) it seems like you're cunningly using this to avoid an automatic scan of all dirs and just telling it where to load the package from. Did I get this right?

If yes, then I assumed that if I'm installing a new package, use-package will install it in the directory specified by `:load-path` but this doesn't seem to be the case and I'm confused about what it actually does. Can you please clarify?

**Comment by thomasf** [comment]
`:load-path` is for locating, not installing. It adds to the emacs `load-path` variable.

From the emacs manual:

```
 -- Variable: load-path
     The value of this variable is a list of directories to search when
     loading files with  load .  Each element is a string (which must be
     a directory name) or  nil  (which stands for the current working
     directory).

```

**Comment by thomasf** [comment]
A simple example

```el
(use-package somepackage
  :load-path "some/directory")
```

Same example but with the macro expanded

```el
(progn
  (eval-and-compile
    (push "/home/a00001/.emacs.d/some/directory" load-path))
  (if
      (not
       (require 'somepackage nil 'noerror))
      (ignore
       (message
        (format "Could not load %s" 'somepackage)))))

```

**Comment by kshenoy** [comment]
> `:load-path` is for locating, not installing. It adds to the emacs load-path variable.

I see. I assumed that if I have something like this:

    (use-package helm
      :load-path "packages/helm"
      :ensure t)

then since use-package is responsible for installing the package, it'll automatically do so to the path specified with `:load-path`. If not, would it make sense to request this feature?

**Comment by vyp** [comment]
> If not, would it make sense to request this feature?

Not to me. I'm not any official collaborator of this project though so just my opinion.

But why would you want that? Packages installed via package.el are normally installed under the `elpa` subdirectory in your emacs user dir I believe (so would be something like `~/.emacs.d/elpa`).

**Comment by kshenoy** [comment]
@vyp Fair enough. I was trying to combine the convenience of auto-install (using `:ensure t`) with the speed of explicitly specifying the `:load-path`. I didn't realize that use-package leverages package.el to do the dirty work.

Ok, then let me rephrase my question as "Is it possible to combine the convenience of auto-install (using `:ensure t`) with the speed of explicitly specifying the `:load-path`?"

**Comment by vyp** [comment]
> Ok, then let me rephrase my question as "Is it possible to combine the convenience of auto-install (using :ensure t) with the speed of explicitly specifying the :load-path?"

First of all, I'm a relatively new user of use-package myself, so I'm not **100%** sure what the purpose of `:load-path` here would be without reading the relevant source code (because I don't need/use it). @jwiegley would be able to answer definitively.

But with that being said, reading the readme:

> If your package needs a directory added to the `load-path` in order to load, use `:load-path`.

It seems fairly straightforward. To rehash it, if your package needs some extra custom `load-path` to load things, you can specify that with the `:load-path` parameter, instead of having to do something like:

``` elisp
(add-to-list 'load-path "~/some/dir")
```

for it in the `:init` or wherever, if I understood correctly. So I believe this is just some syntax sugar (looks nicer, more terse etc).

So I understand what you mean by "the convenience of auto-install (using `:ensure t`)", because yes, it is obviously convenient to tell use-package to handle installation of the package if it's not already installed.

But I do not understand what you mean in the second part of your question "with the speed of explicitly specifying the `:load-path`". What do you mean "with the speed"? What do you mean by "explicitly"? I believe you shouldn't have to explicitly set the load-path for each package if it doesn't require it..

Of course, it's entirely possible that I'm just misunderstanding you here completely, so please don't take my confusion the wrong way. Just trying to help!

> I didn't realize that use-package leverages package.el to do the dirty work.

This is the relevenat section in the readme: https://github.com/jwiegley/use-package#for-packageel-users

**Comment by Silex** [comment]
> I was trying to combine the convenience of auto-install (using :ensure t) with the speed of explicitly specifying the :load-path. I didn't realize that use-package leverages package.el to do the dirty work.

Just curious, what "speed" are you refering to here? Using `:load-path` shouldn't have any effect on how fast a package loads.

**Comment by vyp** [comment]
@Silex that would be @kshenoy, not me. I'm asking that too (with my previous comment).

**Comment by Silex** [comment]
@vyp: yeah, I edited my answer 10 seconds after having posted it

**Comment by thomasf** [comment]
I byte compile the load path in a separate file like thishttps://github.com/thomasf/dotfiles-thomasf-emacs/blob/master/emacs.d/load-path.el the top section of my init.el is also related. I only load package.el at compile time. This also means that no package autoloads are provided you yoi have to specify each one of them you need using :commands

**Comment by thomasf** [comment]
Before this initializing with package.e was more than a third of my emacs start up time. Now it takes less than a second for me to start emacs.

**Comment by Silex** [comment]
@thomasf: hum, why? so it creates less autoloads and thus emacs starts faster?

**Comment by thomasf** [comment]
Yes, but more importaly no scanning of directories, loadig several hundred -autoload.el and maybe -package.el files.. I think the largest issue in play is that emacs cant do this asynchronously.

**Comment by Silex** [comment]
@thomasf: interesting... `use-package` is by default already pretty fast so I didn't think much of improving  load time further (~less than 1 second with 50-100 packages atm)

**Comment by vyp** [comment]
@thomasf Nice. I understand that. But I still don't understand @kshenoy's question? As in, I don't see how `:ensure` and `:load-path` are necessarily related still?

**Comment by Silex** [comment]
@thomasf: I just had this idea: use `use-package` the default way normally, but write a helper function that merges all autoload.el files found in ~/.emacs.d/elpa & prepares the load paths too, which could be automatically run after package-install hooks.

**Comment by thomasf** [comment]
@Silex 

I considered to merge autoloads myself at some point but I decided not to and
do not remember the reason(s) now.

Anyway here are some numbers:

with byte compiled load-path.elc and init.elc:

```
time emacs -l init.elc -batch --eval '(message "Hello, world!")'
Hello, world!

real	0m0.575s
user	0m0.420s
sys     0m0.144s
```


Same but with `(package-initialize)` during start up.

```
time emacs -l init.elc -batch --eval '(message "Hello, world!")'
Hello, world!

real	0m1.433s
user	0m1.184s
sys	0m0.236s

```

Since emacs is running in batch mode in these examples a lot of packages are disabled.


my elpa directory has 417 subdirectories, my init.el has 400 `:ensure` declarations.

The results are when using Emacs's on a low latency 10krpm spinning disk and
after a couple of warm up restarts so that file content can be served from the
disk cache. Apart from the first start up, the difference between ssd and
spinning disks are not noticeable IIRC.

**Comment by thomasf** [comment]
Numbers for full gui startup

without pre compiled load path:

```
time emacs -l init.elc --eval '(kill-emacs)'

real	0m2.846s
user	0m2.336s
sys	0m0.292s
```

and with:

```
time emacs -l init.elc --eval '(kill-emacs)'

real	0m1.771s
user	0m1.304s
sys	0m0.244s
```

I have previously measured that about half of the GUI enacs startup time is
spent setting up fonts and the theme.

**Comment by Silex** [comment]
@thomasf: interesting, thanks for the info

**Comment by kshenoy** [comment]
> @kshenoy: just curious, what "speed" are you refering to here? Using :load-path shouldn't have any effect on how fast a package loads.

@Silex Check out this [video](https://youtu.be/2TSKxxYEbII?list=WL&t=252) in which the author explains how he uses `:load-path` to reduce the startup time.

> As in, I don't see how :ensure and :load-path are necessarily related still?

@vyp They are not. I didn't understand what `:load-path` was meant for and assumed incorrectly that they are.

**Comment by jwiegley** [comment]
`:load-path` is important for `use-package`, instead of just calling `add-to-list` on `load-path`, because of *when* it happens, and how it affects byte-compilation.  Plus, it makes it declaratively clear that you are extending the load path.

For those of us not using `package.el`, or loading in some autoloads file that populates the load-path, this is a highly efficient and package-specific way of extending the load-path.  Plus, if you later disable a package, that path will no longer be on the `load-path`.
